src/main/scala/chisel3/util/random/Lfsr.scala

```diff
@@ -0,0 +1,137 @@
+// See LICENSE for license details.
+
+package chisel3.util.random
+
+import chisel3._
+
+/** Linear Feedback Shift Register (LFSR) generator. This produces a maximal period Galois LFSR of the requested width.
+  * {{{
+  * // Instantiate a new 4-bit LFSR, with the default seed of 1, and automatically stepping.
+  * val lfsr4 = Module(new Lfsr(4))
+  * // Print out the LFSR values in order.
+  * val (count, _) = chisel3.util.Counter(true.B, 16)
+  * printf("  - %d: 0b%b\n", count, input)
+  * // The values this prints are below. The seed (0b0001) is first. The LFSR cycles after 2^4-1 steps.
+  * //  -  0: 0b0001
+  * //  -  1: 0b1100
+  * //  -  2: 0b0110
+  * //  -  3: 0b0011
+  * //  -  4: 0b1101
+  * //  -  5: 0b1010
+  * //  -  6: 0b0101
+  * //  -  7: 0b1110
+  * //  -  8: 0b0111
+  * //  -  9: 0b1111
+  * //  - 10: 0b1011
+  * //  - 11: 0b1001
+  * //  - 12: 0b1000
+  * //  - 13: 0b0100
+  * //  - 14: 0b0010
+  * //  - 15: 0b0001
+  * }}}
+  * @param width LFSR width
+  * @param seed an initial seed (this cannot be zero)
+  */
+class Lfsr(width: Int, seed: BigInt = 1) extends Module {
+  require(seed > 0, "Seed cannot be zero")
+  require(Lfsr.tapsGalois.contains(width), s"No LFSR taps stored for requested size '$width'")
```

Comment:
Is it the case that each sized tap contains only one polynomial (i.e. does it make sense for the user to specify a custom sequence of taps)?

Of course the common case should be easy/fast as it is now.

Reply:
I think there are two use cases: 
1. The current `LFSR16` case of a user wanting a decent pseudo random source. As far as LFSRs go, my understanding is that if they're maximal period then it's good enough. We then provide (hopefully) solutions for all the widths a user cares about. The cited [table](http://courses.cse.tamu.edu/walker/csce680/lfsr_table.pdf) for `2 to 786` which we could scrape and populate here.
2. A user is constructing specific LFSRs for some reason. In that situation, it may make sense to think about an API for constructing general Galois or Fibonacci LFSRs.

I think this PR covers (1) relatively thoroughly and can perhaps be directly improved with more known maximal period LFSR taps.

An API for (2) probably makes sense as well. One thing here is that I think you can only cleanly express a Fibonacci LFSR as a nice polynomial. I.e., `x^4 = x^0 + x^2` is all you need to know for a Fibonacci LFSR and it's correct. All the other `x^3 = x^4, x^2 = x^3, ...` are implicit. For Galois LFSRs you need a sequence of equations to correctly express it. Both are concisely handled as `Set[Int]` passed to a `GaloisLfsr` or `FibonacciLfsr` constructor.

All that said... these are perhaps most cleanly expressed as matrices and it may make sense to just use that under the hood. [The ECC ideas I'm playing around with currently do something like that](https://github.com/freechipsproject/chisel3/pull/1087/files#diff-06af32e6f783e29d75541b98a04471b4R9).

Reply:
In the rebase this has been updated to allow a user to specify custom taps. If the user just wants a maximal period LFSR there are classes/objects to help with that.